metrics-sender/src/main/java/io/knotx/metrics/graphite/GraphiteReporter.java

```diff
@@ -0,0 +1,277 @@
+  private Map<String, Long> lastTimers;
+
+  /**
+   * Creates a new {@link GraphiteReporter} instance.
+   *
+   * @param registry               the {@link MetricRegistry} containing the metrics this
+   *                               reporter will report
+   * @param graphite               the {@link GraphiteSender} which is responsible for sending metrics to a Carbon server
+   *                               via a transport protocol
+   * @param clock                  the instance of the time. Use {@link Clock#defaultClock()} for the default
+   * @param prefix                 the prefix of all metric names (may be null)
+   * @param rateUnit               the time unit of in which rates will be converted
+   * @param durationUnit           the time unit of in which durations will be converted
+   * @param filter                 the filter for which metrics to report
+   * @param executor               the executor to use while scheduling reporting of metrics (may be null).
+   * @param shutdownExecutorOnStop if true, then executor will be stopped in same time with this reporter
+   */
+  public GraphiteReporter(MetricRegistry registry, GraphiteSender graphite, Clock clock,
+      String prefix, TimeUnit rateUnit, TimeUnit durationUnit, MetricFilter filter,
+      ScheduledExecutorService executor, boolean shutdownExecutorOnStop,
+      Set<MetricAttribute> disabledMetricAttributes) {
+    super(registry, graphite, clock, prefix, rateUnit, durationUnit, filter, executor,
+        shutdownExecutorOnStop, disabledMetricAttributes);
+    lastGauges = new HashMap<>();
+    lastCounters = new HashMap<>();
+    lastHistograms = new HashMap<>();
+    lastMeters = new HashMap<>();
+    lastTimers = new HashMap<>();
+  }
+
+  @Override
+  public synchronized void report(@SuppressWarnings("rawtypes") SortedMap<String, Gauge> gauges,
+      SortedMap<String, Counter> counters, SortedMap<String, Histogram> histograms,
+      SortedMap<String, Meter> meters, SortedMap<String, Timer> timers) {
+    super.report(
+        removeRepeating(gauges, lastGauges, Gauge::getValue),
+        removeRepeating(counters, lastCounters, Counter::getCount),
+        removeRepeating(histograms, lastHistograms, Histogram::getCount),
+        removeRepeating(meters, lastMeters, Meter::getCount),
+        removeRepeating(timers, lastTimers, Timer::getCount));
+  }
+
+  protected static <M extends Metric, V> SortedMap<String, M> removeRepeating(
+      SortedMap<String, M> currentMetrics, Map<String, V> previousValues, 
+      Function<M, V> valueFunction) {
+    SortedMap<String, M> result = new TreeMap<>(currentMetrics);
+    Map<String, V> currentValues = currentMetrics.entrySet().stream()
+      .collect(Collectors.toMap(Map.Entry::getKey, e -> valueFunction.apply(e.getValue())));
+    currentValues.entrySet().stream()
+      .filter(e -> previousValues.containsKey(e.getKey()))
+      .filter(e -> Objects.equals(previousValues.get(e.getKey()), e.getValue()))
```

Comment:
Wouldn't that cause some data is not send to the graphite?
I mean, theoretically, some metric like `vertx.http.clients.bloomreach.responses-4xx` could have constant among of `4XX` results in time. Does that filtering prevent sending that information to the metrics db?

Reply:
The metrics sent by Knot.x

Gauge
```
vertx.event-loop-size
vertx.http.clients.connections.max-pool-size
vertx.http.clients.connections.pool-ratio
vertx.pools.worker.vert.x-internal-blocking.max-pool-size
vertx.pools.worker.vert.x-internal-blocking.pool-ratio
vertx.pools.worker.vert.x-worker-thread.max-pool-size
vertx.pools.worker.vert.x-worker-thread.pool-ratio
vertx.worker-pool-size
```

Counter
```
vertx.eventbus.handlers
vertx.eventbus.messages.pending
vertx.eventbus.messages.pending-local
vertx.eventbus.messages.pending-remote
vertx.http.clients.endpoint.<host>:<port>.in-use
vertx.http.clients.endpoint.<host>:<port>.open-netsockets
vertx.http.clients.endpoint.<host>:<port>.queue-size
vertx.http.clients.exceptions
vertx.http.clients.open-connections.<host>
vertx.http.clients.open-netsockets
vertx.http.clients.open-websockets
vertx.pools.worker.vert.x-internal-blocking.in-use
vertx.pools.worker.vert.x-internal-blocking.queue-size
vertx.pools.worker.vert.x-worker-thread.in-use
vertx.pools.worker.vert.x-worker-thread.queue-size
vertx.http.servers.<host>:<port>.exceptions
vertx.http.servers.<host>:<port>.open-netsockets
vertx.http.servers.<host>:<port>.open-websockets
vertx.timers
vertx.verticles
vertx.verticles.<verticle class&gt;
```

Histogram
```
vertx.http.clients.bytes-read
vertx.http.clients.bytes-written
vertx.http.servers.<host>:<port>.bytes-read
vertx.http.servers.<host>:<port>.bytes-written
```

Meter
```
vertx.eventbus.messages.bytes-read
vertx.eventbus.messages.bytes-written
vertx.eventbus.messages.delivered
vertx.eventbus.messages.delivered-local
vertx.eventbus.messages.delivered-remote
vertx.eventbus.messages.published
vertx.eventbus.messages.published-local
vertx.eventbus.messages.published-remote
vertx.eventbus.messages.received
vertx.eventbus.messages.received-local
vertx.eventbus.messages.received-remote
vertx.eventbus.messages.reply-failures
vertx.eventbus.messages.sent
vertx.eventbus.messages.sent-local
vertx.eventbus.messages.sent-remote
vertx.http.clients.responses-1xx
vertx.http.clients.responses-2xx
vertx.http.clients.responses-3xx
vertx.http.clients.responses-4xx
vertx.http.clients.responses-5xx
vertx.http.servers.<host>:<port>.responses-1xx
vertx.http.servers.<host>:<port>.responses-2xx
vertx.http.servers.<host>:<port>.responses-3xx
vertx.http.servers.<host>:<port>.responses-4xx
vertx.http.servers.<host>:<port>.responses-5xx
```

Timer
```
vertx.http.clients.connections
vertx.http.clients.endpoint.<host>:<port>.queue-delay
vertx.http.clients.endpoint.<host>:<port>.ttfb
vertx.http.clients.endpoint.<host>:<port>.usage
vertx.http.clients.requests
vertx.http.clients.<http method>-requests
vertx.http.servers.<host>:<port>.connections
vertx.http.servers.<host>:<port>.requests
vertx.http.servers.<host>:<port>.<http method>-requests
vertx.http.servers.<host>:<port>.<http method>-requests.<endpoint alias>
vertx.pools.worker.vert.x-internal-blocking.queue-delay
vertx.pools.worker.vert.x-internal-blocking.usage
vertx.pools.worker.vert.x-worker-thread.queue-delay
vertx.pools.worker.vert.x-worker-thread.usage
vertx.eventbus.handlers.<handler name>
```

Gauge and Counter are transient values representing current state of a resource.
All the metrics related to the traffic are stored as Histogram, Metric, or Timer. The count value indicates the number of the collected samples, and is always increased.

Reply:
I don't see how that addresses @Skejven's concerns. The fact that some values will survive this code change (due to being cumulative) does not change the fact that this can break some of the metrics, depending on how exactly they're defined.